### PR TITLE
Validate project name as per npm conventions

### LINKED ISSUE #313: Validate project-name for foal createapp
## Present Issue

User can provide any sort of special characters for the projectName with foal createapp which doesn't make any sense.
For instance, `foal createapp !@#` works as per now.

## Proposed solution

Validate the name provided as an argument and show warnings appropriately.

### PR DESCRIPTION
**Pull request by jamesgeorge007**
## Present Issue

> Closes #313

User can provide any sort of special characters for the  `projectName` with `foal createapp` which doesn't make any sense.
For instance, `foal createapp !@#` works as per now.

## Proposed solution

Validate the name provided as an argument and show warnings appropriately.

### CODE REVIEW

File: packages/cli/src/generate/generators/app/create-app.ts [hunk 1]
```diff
@@ -21,6 +21,46 @@ function isYarnInstalled() {
   }
 }
 
+function validateProjectName(name: string) {
+  let flag = 0;
+  const specialChars = ['!',
```
Comment (LoicPoullain) [inline packages/cli/src/generate/generators/app/create-app.ts]: Could you add a comment here with the link of the npm page where the characters were found?

File: packages/cli/src/generate/generators/app/create-app.ts [hunk 2]
```diff
@@ -21,6 +21,46 @@ function isYarnInstalled() {
...
+    ':',
+    ';',
+    '<',
+    '>',
+    ',',
+    '.',
+    '?'
+  ];
+
+  specialChars.map(char => {
```
Comment (LoicPoullain) [inline packages/cli/src/generate/generators/app/create-app.ts]: I think we could use the method `find` here to get a more concise code. Maybe something like `return !specialChars.find(char => name.includes(char))`. This way, we could get rid of the `flag` variable too.

File: packages/cli/src/generate/generators/app/create-app.ts [hunk 3]
```diff
@@ -46,6 +86,14 @@ export async function createApp({ name, sessionSecret, autoInstall }:
     sessionSecret: sessionSecret ? sessionSecret : crypto.randomBytes(16).toString('hex')
   };
 
+  // Validating whether if the project-name follows npm naming conventions
+  if (!validateProjectName(name)) {
+    console.log(
+      red(`\n ${red(`${name} doesn't follow the npm naming conventions. Kindly give a vaild project-name`)}`)
+    );
+    process.exit(1);
```
Comment (LoicPoullain) [inline packages/cli/src/generate/generators/app/create-app.ts]: Using `process.exit` just after a `console.log` may make the message not to be printed in some situations (see https://github.com/FoalTS/foal/pull/309#discussion_r247313980). Could we use a `return` instead? Also, using `process.exit` in the code may make the test process to exit unexpectedly when running the unit tests.

File: packages/cli/src/generate/generators/app/create-app.ts [hunk 4]
```diff
@@ -46,6 +86,14 @@ export async function createApp({ name, sessionSecret, autoInstall }:
     sessionSecret: sessionSecret ? sessionSecret : crypto.randomBytes(16).toString('hex')
   };
 
+  // Validating whether if the project-name follows npm naming conventions
+  if (!validateProjectName(name)) {
```
Comment (LoicPoullain) [inline packages/cli/src/generate/generators/app/create-app.ts]: 👍

**Comment by jamesgeorge007** [comment]
@LoicPoullain There were some terrible merge conflicts which made my working tree dirty. 
I shall open a new PR :+1:

**Comment by LoicPoullain** [comment]
> @LoicPoullain There were some terrible merge conflicts which made my working tree dirty.
> I shall open a new PR 👍

Ok! 👍

**Comment by jamesgeorge007** [comment]
@LoicPoullain  Kindly checkout #326

**Pull request closed without merging**